### PR TITLE
Switch default.qubit to use tensordot

### PR DESCRIPTION
**Pull request by co9olguy**
**Description of the Change:** Modified `default.qubit` plugin to use tensordot for gate applications instead of creating large dense arrays 

**Benefits:** speed improvement of 5-10x

**Possible Drawbacks:** This PR was the biggest bang-for-your-buck change, but code could still be optimized more for speed by using a multi-index representation of states/gates internally. 

**Related GitHub Issues:** none

### CODE REVIEW

File: pennylane/plugins/default_qubit.py [hunk 1]
```diff
@@ -316,11 +316,21 @@ def apply(self, operation, wires, par):
             return
 
         A = self._get_operator_matrix(operation, par)
-
-        # apply unitary operations
-        U = self.expand(A, wires)
-
-        self._state = U @ self._state
+        # TODO: use multi-index vectors/matrices to represent states/gates internally
```
Comment (josh146) [inline pennylane/plugins/default_qubit.py]:
```suggestion

        # TODO: use multi-index vectors/matrices to represent states/gates internally
```

File: pennylane/plugins/default_qubit.py [hunk 2]
```diff
@@ -316,11 +316,21 @@ def apply(self, operation, wires, par):
...
+        A = np.reshape(A, [2] * len(wires) * 2)
+        s = np.reshape(self._state, [2] * self.num_wires)
+        axes = (np.arange(len(wires), 2 * len(wires)), wires)
+        tdot = np.tensordot(A, s, axes=axes)
+        # tensordot causes the axes given in `wires` to end up in the first positions
```
Comment (josh146) [inline pennylane/plugins/default_qubit.py]:
```suggestion

        # tensordot causes the axes given in `wires` to end up in the first positions
```

File: pennylane/plugins/default_qubit.py [hunk 3]
```diff
@@ -316,11 +316,21 @@ def apply(self, operation, wires, par):
...
+        # of the resulting tensor. This corresponds to a (partial) transpose of
+        # the correct output state
+        # We'll need to invert this permutation to put the indices in the correct place
+        unused_idxs = [idx for idx in range(self.num_wires) if idx not in wires]
+        perm = wires + unused_idxs
+        inv_perm = np.argsort(perm) # argsort gives inverse permutation
+        state_multi_index = np.transpose(tdot, inv_perm)
+        state_flat = np.reshape(state_multi_index, 2 ** self.num_wires)
+        self._state = state_flat
```
Comment (josh146) [inline pennylane/plugins/default_qubit.py]: might as well get rid of an extra line, and just have
```python
self._state = np.reshape(state_multi_index, 2 ** self.num_wires)
```

**Comment by co9olguy** [comment]
> Looks great! At the moment, `expand` can't be removed as it's still being used in `self.ev`? It sounds like we could use the exact same line to 'apply' `|psi'> = A|psi>`, and then just take the inner product

Yep, that's the only place it's still used. We can take care of it in a future PR if desired, though this was less of a bottleneck than gate multiplication

File: pennylane/plugins/default_qubit.py (hunk 3 above)
Comment (co9olguy) [reply pennylane/plugins/default_qubit.py]: sounds good, I'll take care of it

Comment (co9olguy) [inline pennylane/plugins/default_qubit.py]:
```suggestion
```

File: pennylane/plugins/default_qubit.py [hunk 4]
```diff
@@ -316,11 +316,21 @@ def apply(self, operation, wires, par):
...
+        tdot = np.tensordot(A, s, axes=axes)
+        # tensordot causes the axes given in `wires` to end up in the first positions
+        # of the resulting tensor. This corresponds to a (partial) transpose of
+        # the correct output state
+        # We'll need to invert this permutation to put the indices in the correct place
+        unused_idxs = [idx for idx in range(self.num_wires) if idx not in wires]
+        perm = wires + unused_idxs
+        inv_perm = np.argsort(perm) # argsort gives inverse permutation
+        state_multi_index = np.transpose(tdot, inv_perm)
+        state_flat = np.reshape(state_multi_index, 2 ** self.num_wires)
```
Comment (co9olguy) [inline pennylane/plugins/default_qubit.py]:
```suggestion
        self._state = np.reshape(state_multi_index, 2 ** self.num_wires)
```

**Pull request merged**